boot/freeldr/freeldr/arch/uefi/uefimem.c

```diff
@@ -256,19 +259,35 @@ UefiMemGetMemoryMap(ULONG *MemoryMapSize)
             }
         }
 
-        UefiSetMemory(FreeldrMem,
-                      MapEntry->PhysicalStart,
-                      MapEntry->NumberOfPages,
-                      MemoryType);
+        /* Sometimes our loader can be loaded into higher memory than we ever allocate */
+        if (MemoryType == LoaderLoadedProgram)
+        {
+            if (((MapEntry->PhysicalStart + (MapEntry->NumberOfPages * PAGE_SIZE)) >> EFI_PAGE_SHIFT) > LoaderPagesSpanned)
+            {
+                /* This value needs to be adjusted if this occurs */
+                LoaderPagesSpanned = ((MapEntry->PhysicalStart + (MapEntry->NumberOfPages * PAGE_SIZE)) >> EFI_PAGE_SHIFT);
+            }
+        }
+
+        /* We really don't want to touch these reserved spots at all */
+        if (MemoryType != LoaderReserve)
+        {
+            UefiSetMemory(FreeldrMem,
+                          MapEntry->PhysicalStart,
+                          MapEntry->NumberOfPages,
+                          MemoryType);
+        }
 
         MapEntry = NEXT_MEMORY_DESCRIPTOR(MapEntry, DescriptorSize);
     }
 
+    /* Reserve the first page, as on some UEFI systems it can be free */
```

Comment:
The comment is not very clear, since a free page is rather should be utilized by the system, than using it as justification for reserving that page as permanently occupied by firmware.

Reply:
We hard coded this is ntoskrnl as a reserved space like windows 2003 does, if we let windows use this as free it’ll assert, both win2k3 and ROS

Reply:
```suggestion
    /* Windows expects the first page to be reserved, otherwise it asserts.
     * However it can be just a free page on some UEFI systems. */
```

Reply:
By the way, I've just seen we have a "ReserveMemory" in uefi. Is it used?

Regarding the first page reservation, this is indeed done in PC-BIOS as well:
https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/freeldr/arch/i386/pc/pcmem.c;hb=a6b281c2280710e684a208deb07abd9838ce7cb6#l654
as well as in PC-98:
https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/freeldr/arch/i386/pc98/pc98mem.c;hb=a6b281c2280710e684a208deb07abd9838ce7cb6#l75
but I wonder where exactly this is done for XBOX (and if not, how ROS can cope with that); I guess it's one of these calls but not sure:
https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/freeldr/arch/i386/xbox/xboxmem.c;hb=a6b281c2280710e684a208deb07abd9838ce7cb6#l250